fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/navbar/EngineViewClippingBehavior.kt

```diff
@@ -0,0 +1,116 @@
+ *
+ * This is useful in combination with [ScrollableToolbar] ensuring the web content is displayed
+ * immediately below / above the toolbar even when it is being animated.
+ *
+ * @param context [Context] used for various Android interactions.
+ * @param attrs XML set attributes configuring this.
+ * @param engineViewParent the parent [View] of the [EngineView].
+ * @param topToolbarHeight size of [ScrollableToolbar] when it is placed above the [EngineView].
+ * We only need the
+ * @param hasTopToolbar whether [EngineView] has a toolbar at the top.
+ */
+class EngineViewClippingBehavior(
+    context: Context?,
+    attrs: AttributeSet?,
+    private val engineViewParent: View,
+    private val topToolbarHeight: Int,
+    private val hasTopToolbar: Boolean,
+) : CoordinatorLayout.Behavior<View>(context, attrs) {
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var engineView = engineViewParent.findViewInHierarchy { it is EngineView } as EngineView?
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var recentBottomToolbarTranslation = 0f
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var recentTopToolbarTranslation = 0f
+
+    override fun layoutDependsOn(parent: CoordinatorLayout, child: View, dependency: View): Boolean {
+        if (dependency is ScrollableToolbar) {
+            return true
+        }
+
+        return super.layoutDependsOn(parent, child, dependency)
+    }
+
+    // If the navbar is enabled, this method will be sequentially called with BrowserToolbar and ToolbarContainerView
+    // as dependencies. This creates a syncing issue because the elements are not aware of each other's translations.
+    // The simplest solution I devised involves making each call adjust the translations for both elements,
+    // while recording the last used values. This approach makes sure that translation won't go async in between calls
+    // and addresses cases where toolbars have different sizes (which they do in the current design).
+    // In such cases, as the top toolbar moves up, the bottom content clipping should be adjusted at twice the speed
+    // to compensate for the increased parent view height.
+    // However, once the top toolbar is completely hidden, the bottom content clipping should move at normal speed.
+    // Access to the latest top toolbar position allows solving this problem through a simple subtraction:
+    // val contentBottomClipping = recentTopToolbarTranslation - recentBottomToolbarTranslation
+    override fun onDependentViewChanged(parent: CoordinatorLayout, child: View, dependency: View): Boolean {
+        // I can't come up with a scenario where the translationY
+        // would return a NaN, but given that it was added to the
+        // previous version of the behaviour and there was a test
+        // written for it, I played it safe.
```

Comment:
Let's link that bug with the prospective fix -> https://bugzilla.mozilla.org/show_bug.cgi?id=1823306

Comment suggestion:

// Added NaN check for translationY as a precaution based on historical issues observed in [linked bug]. This check aims to prevent similar issues, as confirmed by the test. Further investigation might be needed to identify all possible causes of NaN values.